### ISSUE TITLE
Spack install AssertionError assert len(concrete_specs) <= 1

### ISSUE THREAD

**Issue by adamjstewart** [LLNL/spack#2123]
I thought this problem had been solved, but it seems that it's still present (or has regressed) in the latest develop branch. I ran `spack install eigen` and it successfully installed metis. Then it tried to install mvapich2. I hit <kbd>Ctrl</kbd>+<kbd>C</kbd> and killed it because I wanted it to use an already installed version of mvapich2. I eventually gave up and decided to just let Spack build mvapich2 again, but when I reran `spack install eigen`, I got the following error message:

```
$ spack install eigen
==> Installing eigen
==> metis is already installed in /blues/gpfs/home/software/spack-0.9.1/opt/spack/linux-centos6-x86_64/gcc-6.1.0/metis-5.1.0-k4pqzigdgdsuk2r3xls7eyhk6uwvxp46
Traceback (most recent call last):
  File "/soft/spack-0.9.1/bin/spack", line 202, in <module>
    main()
  File "/soft/spack-0.9.1/bin/spack", line 179, in main
    return_val = command(parser, args)
  File "/blues/gpfs/home/software/spack-0.9.1/lib/spack/spack/cmd/install.py", line 117, in install
    package.do_install(**kwargs)
  File "/blues/gpfs/home/software/spack-0.9.1/lib/spack/spack/package.py", line 1183, in do_install
    dirty=dirty)
  File "/blues/gpfs/home/software/spack-0.9.1/lib/spack/spack/package.py", line 1160, in do_install
    rec = spack.installed_db.get_record(self.spec)
  File "/blues/gpfs/home/software/spack-0.9.1/lib/spack/spack/database.py", line 79, in converter
    return function(self, spec_like, *args, **kwargs)
  File "/blues/gpfs/home/software/spack-0.9.1/lib/spack/spack/database.py", line 527, in get_record
    key = self._get_matching_spec_key(spec, **kwargs)
  File "/blues/gpfs/home/software/spack-0.9.1/lib/spack/spack/database.py", line 519, in _get_matching_spec_key
    match = self.query_one(spec, **kwargs)
  File "/blues/gpfs/home/software/spack-0.9.1/lib/spack/spack/database.py", line 653, in query_one
    assert len(concrete_specs) <= 1
AssertionError
```

**NOTE: running `spack reindex` does not fix this problem. I'm going to try manually deleting the metis installation and trying again.**

@tgamblin I would love to see this bug fixed before the 0.10 release.

**Comment by adamjstewart** [comment]
Update: Uninstalling metis and rebuilding it seems to have solved the problem, so if anyone runs into this bug just delete the installation it complains about and try again.

**Comment by pramodk** [comment]
I encountered this issue yesterday! 
Just curious: was `metis` installed in the `past` and you have latest spack while installation of `eigen`?

**Comment by adamjstewart** [comment]
Nope, I had just installed metis minutes before on the same version of Spack. My best guess as to the cause of the problem is that Spack is supposed to add each successful installation to the database as soon as it is installed, but it is actually waiting until the full requested DAG is built before adding the installations to the database.

**Comment by tgamblin** [comment]
@adamjstewart: are the dependency installs not showing up in the database?  Spack should not (as of #1562) wait until the full DAG is built to update the DB -- it locks the DB once per package install to write it.

I believe that this issue is related to a number of the issues in the 0.10 project.  Specifically the ones about matching by hash and the one about Spack treating unknown OS's as `None`.

I *believe* you see this particular error because you have an two installs of `metis` that differ only by OS, and Spack treats `None` as unconstrained when it compares specs, so they match each other when they should not.  There are really two issues here:
  1. Concrete specs should just match by hash when you do `satisfies()`, and skip the complicated stuff.  I have a branch with this fix that I am testing out.  This fixes issues with uninstall by hash and other places where specs appear to be ambiguous when they should not.
  1. Concrete

**Comment by adamjstewart** [comment]
> I _believe_ you see this particular error because you have an two installs of `metis` that differ only by OS, and Spack treats `None` as unconstrained when it compares specs, so they match each other when they should not. 

I don't believe that this is the case. All of my installations are under `linux-centos6-x86_64` or `linux-redhat5-x86_64`, and there are only metis installations in the centos6 directory. I do have 4 different installations of metis, but they only differ by compiler and compiler-version.

**Comment by tgamblin** [comment]
Ok I will have to look at this in more detail then.

**Comment by tgamblin** [comment]
@adamjstewart: does metis not appear in the DB if you ctrl-c after it is installed?

**Comment by adamjstewart** [comment]
How can I tell?

**Comment by tgamblin** [comment]
1. Do `spack install eigen`, and let it build and install `metis` (I'm assuming metis is not installed)
1. Hit `ctrl-C` once you see the `[+]` line for `metis`
1. Do a `spack find` and check whether the just-installed `metis` appears in the output.

**Comment by adamjstewart** [comment]
Hmm, I'm having trouble reproducing this problem. I have metis installed now but I don't want to uninstall it because I just told a user eigen is installed. I wasn't able to reproduce the problem on a different machine either. @pramodk Do you have any other details on how the problem happened for you?

**Comment by pramodk** [comment]
@adamjstewart : I am also building only on `linux`. While installing some packages I did `Ctl-C` few times. (just got this issue again while sending your response in #2130).

**Comment by adamjstewart** [comment]
Does `spack find` find the package in the database? What is in the installation directory and `.spack` directory?

**Comment by pramodk** [comment]
I see following:

```bash
spack install -v mypkg ^hdf5+mpi%gcc
==> Installing mypkg
==> mvapich2 is externally installed in /somepath/mvapich2/mvapich2-2.0.1-nocuda-slurm-14.03.4.2/install
==> cmake is externally installed in cmake-3.6.1-gcc-4.4.7-xhpqagp
==> hdf5 is already installed in /pathxx/spack/opt/spack/linux-redhat6-x86_64/gcc-4.9.0/hdf5-1.10.0-patch1-gkx34kcxlc7yyy75zhgbdhtuodmm65pp
Traceback (most recent call last):
  File "/pathxx/spack/bin/spack", line 202, in <module>
    main()
  File "/pathxx/spack/bin/spack", line 179, in main
    return_val = command(parser, args)
  File "/pathxx/spack/lib/spack/spack/cmd/install.py", line 117, in install
    package.do_install(**kwargs)
  File "/pathxx/spack/lib/spack/spack/package.py", line 1183, in do_install
    dirty=dirty)
  File "/pathxx/spack/lib/spack/spack/package.py", line 1160, in do_install
    rec = spack.installed_db.get_record(self.spec)
  File "/pathxx/spack/lib/spack/spack/database.py", line 79, in converter
    return function(self, spec_like, *args, **kwargs)
  File "/pathxx/spack/lib/spack/spack/database.py", line 527, in get_record
    key = self._get_matching_spec_key(spec, **kwargs)
  File "/pathxx/spack/lib/spack/spack/database.py", line 519, in _get_matching_spec_key
    match = self.query_one(spec, **kwargs)
  File "/pathxx/spack/lib/spack/spack/database.py", line 653, in query_one
    assert len(concrete_specs) <= 1
AssertionError
```
you mean that last `hdf5` package? yes, `spack find` shows `hdf5` package with hash `gkx34kcxlc7yyy75zhgbdhtuodmm65pp`.

**Comment by pramodk** [comment]
@adamjstewart @tgamblin : today I saw this error may be 7-8 times! The pattern I am seeing is that when I install a package with different compiler flags and try to install  it again with another flags then saw this error: 

```bash
spack find -v -f mypkg %pgi    #installed package with twi different flags

==> 2 installed packages.
-- linux-redhat6-x86_64 / pgi@16.5-0 ----------------------------
mypkg@develop%pgi
mypkg@develop%pgi cflags="-O3 -fPIC" cppflags="-O3 -fPIC" cxxflags="-O3 -fPIC"

#trying to install it again
→ spack install --dirty -v mypkg %pgi ^mpich cflags=\"-O1 -g\" cxxflags=\"-O1 -g\"
==> mypkg is already installed in /somepath/spack/opt/spack/linux-redhat6-x86_64/pgi-16.5-0/mypkg-develop-ww2orjtn6eux7tsnkcznp6aebwfssz2w
Traceback (most recent call last):
  File "/somepath/spack/bin/spack", line 192, in <module>
    main()
  File "/somepath/spack/bin/spack", line 169, in main
    return_val = command(parser, args)
  ……
  File "/somepath/spack/lib/spack/spack/database.py", line 519, in _get_matching_spec_key
    match = self.query_one(spec, **kwargs)
  File "/somepath/spack/lib/spack/spack/database.py", line 653, in query_one
    assert len(concrete_specs) <= 1
AssertionError
```
Does this help?

**Comment by pramodk** [comment]
And trying to uninstall by hash gave me:

```bash
spack uninstall /ww2orjtn6eux7tsnkcznp6aebwfssz2w
==> Error: mypkg@develop%pgi@16.5-0 cflags="-O3 -fPIC" cppflags="-O3 -fPIC" cxxflags="-O3 -fPIC" arch=linux-redhat6-x86_64^mpich@3.2%pgi@16.5-0 cflags="-O3 -fPIC" cppflags="-O3 -fPIC" cxxflags="-O3 -fPIC"+hydra+pmi+romio~verbs arch=linux-redhat6-x86_64 matches multiple packages:

-- linux-redhat6-x86_64 / pgi@16.5-0 ----------------------------
2bw2hr6 mypkg@develop%pgi
ww2orjt mypkg@develop%pgi cflags="-O3 -fPIC" cppflags="-O3 -fPIC" cxxflags="-O3 -fPIC"
```

**Comment by adamjstewart** [comment]
@tgamblin I'm seeing this error message again on the latest develop. I'm trying to install mvapich2 with an old already-installed libpciaccess. Libpciaccess is in the database, but it's still crashing anyway. I don't think I can delete this one because several other installed packages depend on it. I'm getting ready to install a third Spack repository, but I was hoping I could wait until 0.10 was out before having to do that again. Any suggestions? If the Spack database gets corrupted or the specs change this frequently, using Spack in production will be a nightmare.

I tried creating a debug tarball, but I'm getting this error message. Is this a known bug? It might be related to the new config.yaml support:
```
$ spack debug create-db-tarball
Traceback (most recent call last):
  File "/soft/spack-0.9.1/bin/spack", line 203, in <module>
    main()
  File "/soft/spack-0.9.1/bin/spack", line 180, in main
    return_val = command(parser, args)
  File "/blues/gpfs/home/software/spack-0.9.1/lib/spack/spack/cmd/debug.py", line 100, in debug
    action[args.debug_command](args)
  File "/blues/gpfs/home/software/spack-0.9.1/lib/spack/spack/cmd/debug.py", line 77, in create_db_tarball
    base = os.path.basename(spack.install_path)
AttributeError: 'module' object has no attribute 'install_path'
```

**Comment by adamjstewart** [comment]
Ok, this is getting absurd. I tried installing again with a different compiler and everything built fine, but the package I'm debugging crashed. So I ran `spack clean` and `spack install` again, and now I'm seeing this error message for mvapich2 again. This time I guess I can uninstall it and try again, but I would really love help with this.

**Comment by adamjstewart** [comment]
Ha, I forgot identifying specs by hash was broken too:
```
$ spack uninstall /v6xdy4ccf7drmudbjkxa2vu6ctovy2jj
==> Error: mvapich2@2.2%gcc@5.3.0~debug~gforker+hydra~mrail~nemesis~nemesisib~nemesisibtcp+psm~remshell~slurm~sock arch=linux-centos6-x86_64^bison@3.0.4%gcc@5.3.0 arch=linux-centos6-x86_64^libpciaccess@0.13.4%gcc@5.3.0 arch=linux-centos6-x86_64 matches multiple packages:

-- linux-centos6-x86_64 / gcc@5.3.0 -----------------------------
v6xdy4c mvapich2@2.2%gcc~debug~gforker+hydra~mrail~nemesis~nemesisib~nemesisibtcp+psm~remshell~slurm~sock

m6iglhi mvapich2@2.2b%gcc~debug~gforker+hydra~mrail~nemesis~nemesisib~nemesisibtcp+psm~remshell~slurm~sock


==> Error: You can either:
    a) Use a more specific spec, or
    b) use spack uninstall -a to uninstall ALL matching specs.
```

**Comment by adamjstewart** [comment]
Hmm, but this shouldn't be happening:
```
$ spack uninstall mvapich2@2.2 %gcc@5
==> Error: mvapich2@2.2%gcc@5 matches multiple packages:

-- linux-centos6-x86_64 / gcc@5.3.0 -----------------------------
v6xdy4c mvapich2@2.2%gcc~debug~gforker+hydra~mrail~nemesis~nemesisib~nemesisibtcp+psm~remshell~slurm~sock

m6iglhi mvapich2@2.2b%gcc~debug~gforker+hydra~mrail~nemesis~nemesisib~nemesisibtcp+psm~remshell~slurm~sock


==> Error: You can either:
    a) Use a more specific spec, or
    b) use spack uninstall -a to uninstall ALL matching specs.
```
One is version 2.2, the other is 2.2b, but Spack thinks they are the same version, or at least it thinks that 2.2 satisfies 2.2 and 2.2b...

**Comment by adamjstewart** [comment]
And... manually deleting the mvapich2 installation results in another libpciaccess bug. I give up.

**Comment by adamjstewart** [comment]
Had this happen to me again today. I tried installing hdf and hdf5, but they both crashed. This obviously led to zlib being installed. Then, I successfully build hdf by adding a compiler flag. This caused another zlib installation with the compiler flag as the only difference. Then I tried installing hdf5 with the same compiler flag and I got:
```
==> Installing hdf5
==> zlib is already installed in /blues/gpfs/home/software/spack-0.9.1/opt/spack/linux-centos6-x86_64/nag-6.1/zlib-1.2.8-csnxjvw4tsixcouvkruxgrv6dfbncjbh
Traceback (most recent call last):
  File "/soft/spack-0.9.1/bin/spack", line 203, in <module>
    main()
  File "/soft/spack-0.9.1/bin/spack", line 180, in main
    return_val = command(parser, args)
  File "/blues/gpfs/home/software/spack-0.9.1/lib/spack/spack/cmd/install.py", line 342, in install
    package.do_install(**kwargs)
  File "/blues/gpfs/home/software/spack-0.9.1/lib/spack/spack/package.py", line 1189, in do_install
    **kwargs
  File "/blues/gpfs/home/software/spack-0.9.1/lib/spack/spack/package.py", line 1161, in do_install
    rec = spack.store.db.get_record(self.spec)
  File "/blues/gpfs/home/software/spack-0.9.1/lib/spack/spack/database.py", line 80, in converter
    return function(self, spec_like, *args, **kwargs)
  File "/blues/gpfs/home/software/spack-0.9.1/lib/spack/spack/database.py", line 529, in get_record
    key = self._get_matching_spec_key(spec, **kwargs)
  File "/blues/gpfs/home/software/spack-0.9.1/lib/spack/spack/database.py", line 521, in _get_matching_spec_key
    match = self.query_one(spec, **kwargs)
  File "/blues/gpfs/home/software/spack-0.9.1/lib/spack/spack/database.py", line 655, in query_one
    assert len(concrete_specs) <= 1
AssertionError
```
This time I was able to create a debug tarball. Here it is:
[spack-db.develop.1eeef97.2016-11-03-142458.tar.gz](https://github.com/LLNL/spack/files/569923/spack-db.develop.1eeef97.2016-11-03-142458.tar.gz)

**Comment by tgamblin** [comment]
@adamjstewart: can you test with #2227?

**Comment by adamjstewart** [comment]
Testing isn't easy because I still haven't found a way to reproduce it with 100% certainty. I can let you know if it happens again though. @pramodk Can you reproduce it 100% of the time when using different CFLAGS?

**Comment by pramodk** [comment]
> Can you reproduce it 100% of the time when using different CFLAGS?

I was able to reproduce this with `cflags` and don't see it now. But I was getting this error before I even started using `cflags / cxxflags`. So I am not sure #2227 is the fix for all scenarios. But I also have to wait until I hit this error again.

**Comment by adamjstewart** [comment]
Just saw this error message again.

**Comment by adamjstewart** [comment]
Also, I'm seeing duplicate packages for all of my installations. Did the hashing mechanism change again??

**Comment by adamjstewart** [comment]
Happened again to me today.

**Comment by adamjstewart** [comment]
Ran into this problem again today... and found another clue! I'm working on a new package, which only has a single dependency, pkg-config (build deptype). When I ran `spack install`, it found that pkg-config was already installed, but then crashed with the error message I've seen dozens of times before. Since the package only had a single dependency, I tried uninstalling it, only to find:
```
$ spack uninstall -a pkg-config%gcc
==> The following packages will be uninstalled : 

-- linux-centos6-x86_64 / gcc@6.1.0 -----------------------------
3e42kgk pkg-config@0.29.1%gcc+internal_glib

hcs5k2j pkg-config@0.29.1%gcc+internal_glib


==> Do you want to proceed ? [y/n]
y
==> Successfully uninstalled pkg-config@0.29.1%gcc@6.1.0+internal_glib arch=linux-centos6-x86_64-3e42kgk
==> Removed stale DB entry for pkg-config@0.29.1%gcc@6.1.0+internal_glib arch=linux-centos6-x86_64-3e42kgk
```
A couple of things to note here. Spack lists two identical installations with different hashes. When I uninstall all, it only mentions uninstalling one of the hashes. The "Removed stale DB entry" is definitely an important clue. Anyone know where that might come from? @tgamblin

**Comment by becker33** [comment]
Can you add the output of a later `spack find -l` to this?

**Comment by adamjstewart** [comment]
After the uninstall and after reinstalling, I see:
```
$ spack find -l pkg-config%gcc
==> 1 installed packages.
-- linux-centos6-x86_64 / gcc@6.1.0 -----------------------------
3e42kgk pkg-config@0.29.1
```
So `hcs5k2j` isn't still hanging around. And yes, it reinstalled the exact same hash.

**Comment by tgamblin** [comment]
@adamjstewart: should be fixed by #2626.  Can you verify?

**Comment by adamjstewart** [comment]
@tgamblin This might actually be finally fixed! I ran into the same problem again when trying to build mvapich2, updated to the latest develop, tried again, and it worked!
